pkg/controller/virtualmachine/virtualmachine_controller.go

```diff
@@ -92,50 +92,58 @@ func (r *ReconcilePolicy) Reconcile(request reconcile.Request) (reconcile.Result
 		// The object is not being deleted, so if it does not have a finalizer,
 		// then lets add the finalizer and update the object.
-		err = r.addFinalizerAndUpdate(instance, &request)
+		err = r.addFinalizerAndUpdate(instance, &request, podFinalizerName)
 		return reconcile.Result{}, err
 	}
 	// The object is being deleted
 	log.V(1).Info("The VM is being marked for deletion",
 		"virtualMachineName", request.Name,
 		"virtualMachineNamespace", request.Namespace)
-	err = r.removeFinalizerAndReleaseMac(instance, &request)
+	err = r.removeFinalizerAndReleaseMac(instance, &request, podFinalizerName)
 	return reconcile.Result{}, err
 }
 
-func (r *ReconcilePolicy) addFinalizerAndUpdate(virtualMachine *kubevirt.VirtualMachine, request *reconcile.Request) error {
-	if helper.ContainsString(virtualMachine.ObjectMeta.Finalizers, pool_manager.RuntimeObjectFinalizerName) {
+func (r *ReconcilePolicy) addFinalizerAndUpdate(virtualMachine *kubevirt.VirtualMachine, request *reconcile.Request, podFinalizerName string) error {
+	if helper.ContainsString(virtualMachine.ObjectMeta.Finalizers, podFinalizerName) {
 		return nil
 	}
 
 	log.V(1).Info("The VM does not have a finalizer",
 		"virtualMachineName", request.Name,
-		"virtualMachineNamespace", request.Namespace)
-	virtualMachine.ObjectMeta.Finalizers = append(virtualMachine.ObjectMeta.Finalizers, pool_manager.RuntimeObjectFinalizerName)
+		"virtualMachineNamespace", request.Namespace,
+		"podFinalizerName", podFinalizerName)
 
-	err := r.Update(context.Background(), virtualMachine)
-	if err != nil {
-		log.Error(err, "failed to update the VM with the new finalizer",
+	err, isMacAllocatedToRange := r.poolManager.MarkVMAsReady(virtualMachine)
+
+	//we want to add a finalizer only if this pod has allocated a mac to this vm
+	if isMacAllocatedToRange {
+		virtualMachine.ObjectMeta.Finalizers = append(virtualMachine.ObjectMeta.Finalizers, podFinalizerName)
+
+		err = r.Update(context.Background(), virtualMachine)
+		if err != nil {
+			log.Error(err, "failed to update the VM with the new finalizer",
+				"virtualMachineName", request.Name,
+				"virtualMachineNamespace", request.Namespace)
+			return err
+		}
+
+		log.V(1).Info("Finalizer was added to the VM",
 			"virtualMachineName", request.Name,
 			"virtualMachineNamespace", request.Namespace)
-		return err
 	}
+	return err
 
-	log.V(1).Info("Finalizer was added to the VM",
-		"virtualMachineName", request.Name,
-		"virtualMachineNamespace", request.Namespace)
-
-	return r.poolManager.MarkVMAsReady(virtualMachine)
 }
 
-func (r *ReconcilePolicy) removeFinalizerAndReleaseMac(virtualMachine *kubevirt.VirtualMachine, request *reconcile.Request) error {
-	if !helper.ContainsString(virtualMachine.ObjectMeta.Finalizers, pool_manager.RuntimeObjectFinalizerName) {
+func (r *ReconcilePolicy) removeFinalizerAndReleaseMac(virtualMachine *kubevirt.VirtualMachine, request *reconcile.Request, podFinalizerName string) error {
```

Comment:
ditto

except here the logic of finalizer handling and poolManager is already mixed.